src/thread_task.rs

```diff
@@ -834,163 +835,151 @@ pub unsafe extern "C" fn rav1d_worker_task(data: *mut c_void) -> *mut c_void {
                                 let start = frame_hdr.tiling.row_start_sb[ntr as usize] as c_int;
                                 if (*next_t).sby == start {
-                                    (*f).frame_thread.next_tile_row[p as usize] = ntr;
+                                    f.frame_thread.next_tile_row[p as usize] = ntr;
                                 }
                                 (*next_t).recon_progress = (*next_t).sby + 1;
                                 insert_task(f, next_t, 0 as c_int);
                             }
-                            found = true;
-                            break 'decoding;
-                        } else if (*t).type_0 as c_uint == RAV1D_TASK_TYPE_CDEF as c_int as c_uint {
-                            let p1_1 = (*f).frame_thread.copy_lpf_progress
-                                [((*t).sby - 1 >> 5) as usize]
+                            break 'found (f, t, prev_t);
+                        } else if t.type_0 as c_uint == RAV1D_TASK_TYPE_CDEF as c_int as c_uint {
+                            let p1_1 = f.frame_thread.copy_lpf_progress[(t.sby - 1 >> 5) as usize]
                                 .load(Ordering::SeqCst);
-                            if p1_1 as c_uint & (1 as c_uint) << ((*t).sby - 1 & 31) != 0 {
-                                found = true;
-                                break 'decoding;
+                            if p1_1 as c_uint & (1 as c_uint) << (t.sby - 1 & 31) != 0 {
+                                break 'found (f, t, prev_t);
                             }
                         } else {
-                            if (*t).deblock_progress == 0 {
+                            if t.deblock_progress == 0 {
                                 unreachable!();
                             }
-                            let p1_2 = (*f).frame_thread.deblock_progress.load(Ordering::SeqCst);
-                            if p1_2 >= (*t).deblock_progress {
-                                (*f).task_thread
+                            let p1_2 = f.frame_thread.deblock_progress.load(Ordering::SeqCst);
+                            if p1_2 >= t.deblock_progress {
+                                f.task_thread
                                     .error
                                     .fetch_or((p1_2 == TILE_ERROR) as c_int, Ordering::SeqCst);
-                                found = true;
-                                break 'decoding;
+                                break 'found (f, t, prev_t);
                             }
                         }
                     }
                     // next:
-                    prev_t = t;
-                    t = (*t).next;
-                    (*f).task_thread.task_cur_prev = prev_t;
+                    next_t = t.next.as_mut();
+                    prev_t = Some(t);
+                    f.task_thread.task_cur_prev = prev_t.as_deref_mut().unwrap() as *mut Rav1dTask;
```

Comment:
Can't reborrow it because we moved it into prev_t before that. Could work around this but I don't think it's worth it.

Reply:
Ohh.  But the only reason it works this way is because `f.task_thread.task_cur_prev` is a `*mut`.  You'll have multiple `&mut`s to `prev_t` once that's cleaned up.  I think you can just replace uses of `prev_t` with `f.task_thread.task_cur_prev.as_mut()`.  The problem with that is then you're `break 'found (f, t, f.task_thread.task_cur_prev.as_mut()`, and the `f` can also access `f.task_thread.task_cur_prev`.  That's still a problem with your current approach; it's just hidden by the raw pointer.  One way of getting around that is only doing `break 'found (f, t)` everywhere and then `let mut prev_t = f.task_thread.task_cur_prev.as_mut();` after.  And where you did `break 'found (f, t, None)` before, you can just `assert!(f.task_thread.task_cur_prev.is_null())` beforehand, which works in the tests.

If you'd like, you could fix this in a follow-up PR, but it seems like it's gonna need fixing either way.

Reply:
I actually want to replace the entire task list with something else, because it's a linked list right now and that's a pain in Rust. We might be able to keep the linked list with Arcs, but that needs further thought and investigation. I'm planning to look into this. We shouldn't do anything extra here until that gets figured out.

Reply:
Okay that's fine then if you plan to replace it and make it safe some other way.